File: tower-tussle/android/app/src/main/java/com/dabit3/towertussle/PlayerProfile.kt
package com.dabit3.towertussle

import android.content.Context
import android.content.SharedPreferences
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.setValue
import kotlin.math.max

class PlayerProfile(context: Context) {
    private object Keys {
        const val TROPHIES = "tt.trophies"
        const val GOLD = "tt.gold"
        const val WINS = "tt.wins"
        const val LOSSES = "tt.losses"
        const val DRAWS = "tt.draws"
        const val DECK = "tt.deck"
    }

    private val prefs: SharedPreferences = context.getSharedPreferences("tower_tussle", Context.MODE_PRIVATE)

    var trophies by mutableIntStateOf(prefs.getInt(Keys.TROPHIES, 0)); private set
    var gold by mutableIntStateOf(prefs.getInt(Keys.GOLD, 100)); private set
    var wins by mutableIntStateOf(prefs.getInt(Keys.WINS, 0)); private set
    var losses by mutableIntStateOf(prefs.getInt(Keys.LOSSES, 0)); private set
    var draws by mutableIntStateOf(prefs.getInt(Keys.DRAWS, 0)); private set
    val deck = mutableStateListOf<String>()

    init {
        val saved = prefs.getString(Keys.DECK, null)?.split(",")?.filter { it.isNotBlank() } ?: emptyList()
        val valid = saved.size == 8 && saved.all { id -> Cards.all.any { it.id == id } }
        deck.addAll(if (valid) saved else Cards.defaultDeck)
    }

    val collection: List<String> get() = Cards.all.map { it.id }.filter { it !in deck }

    val averageElixir: Double get() = deck.sumOf { Cards.byId(it).cost }.toDouble() / max(deck.size, 1)

    fun apply(result: MatchResult) {
        trophies = max(0, trophies + result.trophyDelta)
        gold += result.goldDelta
        when (result.outcome) {
            MatchOutcome.VICTORY -> wins++
            MatchOutcome.DEFEAT -> losses++
            MatchOutcome.DRAW -> draws++
        }
        save()
    }

    fun swap(deckCard: String, collectionCard: String) {
        val i = deck.indexOf(deckCard)
        if (i < 0 || collectionCard !in collection) return
        deck[i] = collectionCard
        save()
    }

    fun resetDeck() {
        deck.clear()
        deck.addAll(Cards.defaultDeck)
        save()
    }

    private fun save() {
        prefs.edit()
            .putInt(Keys.TROPHIES, trophies)
            .putInt(Keys.GOLD, gold)
            .putInt(Keys.WINS, wins)
            .putInt(Keys.LOSSES, losses)
            .putInt(Keys.DRAWS, draws)
            .putString(Keys.DECK, deck.joinToString(","))
            .apply()
    }
}

File: tower-tussle/ios/TowerTussle/PlayerProfile.swift
import Foundation
import Combine

final class PlayerProfile: ObservableObject {
    private enum Keys {
        static let trophies = "tt.trophies"
        static let gold = "tt.gold"
        static let wins = "tt.wins"
        static let losses = "tt.losses"
        static let draws = "tt.draws"
        static let deck = "tt.deck"
    }

    private let defaults: UserDefaults

    @Published private(set) var trophies: Int
    @Published private(set) var gold: Int
    @Published private(set) var wins: Int
    @Published private(set) var losses: Int
    @Published private(set) var draws: Int
    @Published private(set) var deck: [String]

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        trophies = defaults.integer(forKey: Keys.trophies)
        gold = defaults.object(forKey: Keys.gold) as? Int ?? 100
        wins = defaults.integer(forKey: Keys.wins)
        losses = defaults.integer(forKey: Keys.losses)
        draws = defaults.integer(forKey: Keys.draws)
        let saved = defaults.stringArray(forKey: Keys.deck) ?? []
        deck = saved.count == 8 && saved.allSatisfy({ id in Cards.all.contains { $0.id == id } }) ? saved : Cards.defaultDeck
    }

    var collection: [String] { Cards.all.map(\.id).filter { !deck.contains($0) } }

    var averageElixir: Double {
        Double(deck.map { Cards.byId($0).cost }.reduce(0, +)) / Double(max(deck.count, 1))
    }

    func apply(_ result: MatchResult) {
        trophies = max(0, trophies + result.trophyDelta)
        gold += result.goldDelta
        switch result.outcome {
        case .victory: wins += 1
        case .defeat: losses += 1
        case .draw: draws += 1
        }
        save()
    }

    func swap(deckCard: String, with collectionCard: String) {
        guard let i = deck.firstIndex(of: deckCard), collection.contains(collectionCard) else { return }
        deck[i] = collectionCard
        save()
    }

    func resetDeck() {
        deck = Cards.defaultDeck
        save()
    }

    private func save() {
        defaults.set(trophies, forKey: Keys.trophies)
        defaults.set(gold, forKey: Keys.gold)
        defaults.set(wins, forKey: Keys.wins)
        defaults.set(losses, forKey: Keys.losses)
        defaults.set(draws, forKey: Keys.draws)
        defaults.set(deck, forKey: Keys.deck)
    }
}
